### PR TITLE
Argo CD with CodeCommit Contribution

### LINKED ISSUE #367: [Automation Lab] ArgoCD Tests
### What would you like to be added?

In this PR : https://github.com/aws-samples/eks-workshop-v2/pull/354 we added a new module for ArgoCD.

The modules tests are not written yet.

### Why is this needed?

This Issue is for tracking that we correctly test this module according to : https://github.com/aws-samples/eks-workshop-v2/blob/main/docs/automated_tests.md

### LINKED ISSUE #376: [feat] Add ArgoCD lab with AWS CodeCommit under Automation
### What would you like to be added?

A an [ArgoCD](https://argo-cd.readthedocs.io/en/stable/) lab to the [Automation](https://www.eksworkshop.com/docs/automation) under the subsection [GitOps](https://www.eksworkshop.com/docs/automation/gitops/)

ArgoCD should follow similar scenario as the Flux lab, meaning using kustomize for example and use AWS Codecommit as a source repository

### Why is this needed?

To show example of Argo CD GitOps with AWS Codecommit in addition to implementation with GitHub from #354

### LINKED ISSUE #392: Upgrade ArgoCD version for ArgoCD lab
### What would you like to be added?

Upgrade the version of ArgoCD helm chart used in terraform eks addon for the ArgoCD lab

### Why is this needed?

Keep up to date with bug and security fixes

### PR DESCRIPTION
**Pull request by ybezsonov**
#### Which issue(s) this PR fixes:

Fixes #376 
Fixes #392 
Fixes #367

#### Quality checks

- [x] My content adheres to the style guidelines
- [x] I ran `make test` or `make e2e-test` and it was successful

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: website/docs/automation/gitops/argocd-codecommit/cluster_bootstrap.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+Let's break down the manifest above:
+
+- First we tell ArgoCD what will be the `name` of our application
+- After that we define which Git repository `source` to use to get the desired state of our application. We define `repoUrl` of the AWS CodeCommit repository, a repository branch with `targetRevision`
+- Finally we define a `path` to application manifests within the repository
+
+Let's apply the manifest:
+
+```bash
+$ envsubst < /workspace/modules/automation/gitops/argocd-codecommit/argocd-application.yaml | kubectl apply -f -
```
Comment (niallthomson) [inline website/docs/automation/gitops/argocd-codecommit/cluster_bootstrap.md:23]: Lets change this to use kustomize to substitute in the environment variables to make it consistent with all our other labs. I would suggest moving the Argo application in to its own directory (like `environment/workspace/modules/automation/gitops/argocd-codecommit/application`) then apply it with `kubectl apply -k`.

You can see an example of us doing this with a security group policy here:

https://github.com/aws-samples/eks-workshop-v2/tree/main/environment/workspace/modules/networking/securitygroups-for-pods/sg

Please also add the Argo Application CRD to the end of this list so it gets cleaned up by our reset-environment script https://github.com/aws-samples/eks-workshop-v2/blob/main/environment/bin/reset-environment#L43

File: terraform/modules/cluster/addons.tf [hunk 2]
```diff
@@ -236,6 +237,39 @@ module "eks_blueprints_kubernetes_addons" {
     additional_irsa_policies = ["arn:aws:iam::aws:policy/AdministratorAccess"]
   }
 
+  argocd_helm_config = {
```
Comment (niallthomson) [inline terraform/modules/cluster/addons.tf:240]: Can we make sure the Argo components run on the system node group? See this example where we do it for the load balancer controller https://github.com/aws-samples/eks-workshop-v2/blob/main/terraform/modules/cluster/addons.tf#L102

File: website/docs/automation/gitops/argocd-codecommit/cleanup.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+title: "Cleanup"
+sidebar_position: 90
+---
+
+To remove Argo CD configuration for `apps` from the cluster run:
+
+```bash
+$ kubectl -n argocd delete secret codecommit-repo
+secret "codecommit-repo" deleted
+$ kubectl -n argocd delete application apps
```
Comment (niallthomson) [inline website/docs/automation/gitops/argocd-codecommit/cleanup.md:11]: You'll can remove this once you update the reset-environment script as outlined somewhere else in this review.

File: website/docs/automation/gitops/argocd-codecommit/deploy_application.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+```bash
+$ echo "ArgoCD URL: http://$(kubectl get svc argo-cd-argocd-server -n argocd -o json | jq --raw-output '.status.loadBalancer.ingress[0].hostname')"
+$ echo "ArgoCD admin password: $(kubectl -n argocd get secret argocd-initial-admin-secret -o jsonpath="{.data.password}" | base64 -d)"
+```
+
+:::info
+
+If you need to reset Argo CD `admin` password, the sequence of commands below can be used:
+
+kubectl -n argocd patch secret argocd-secret -p '{"data": {"admin.password": null, "admin.passwordMtime": null}}'
```
Comment (niallthomson) [inline website/docs/automation/gitops/argocd-codecommit/deploy_application.md:85]: You should be able to use a ```bash``` section to wrap these commands so they're a bit easier to see.

**Review by niallthomson — CHANGES_REQUESTED** [review]
Overall a great foundation for an ArgoCD lab. Appreciate the work put in here and the effort made to align with the other labs. I've made some comments around some of the manifests and Terraform.

File: terraform/modules/cluster/codecommit.tf [hunk 5]
```diff
@@ -3,6 +3,11 @@ resource "aws_codecommit_repository" "gitops" {
   description     = "CodeCommit repository for GitOps"
 }
 
+resource "aws_codecommit_repository" "gitops-argocd" {
```
[the remark this answers is not in the document]
Comment (ybezsonov) [reply terraform/modules/cluster/codecommit.tf]: fixed

File: terraform/modules/cluster/addons.tf (hunk 2 above)
Comment (ybezsonov) [reply terraform/modules/cluster/addons.tf]: Used approach from PR 354

File: website/docs/automation/gitops/argocd-codecommit/cluster_bootstrap.md (hunk 1 above)
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd-codecommit/cluster_bootstrap.md]: Created separate folder for Argo CD kustomization and changed to solution with "replacements" which should be future proof.

Comment (ybezsonov) [reply website/docs/automation/gitops/argocd-codecommit/cluster_bootstrap.md]: Addedd Argo CD CRD to reset-environment

File: website/docs/automation/gitops/argocd-codecommit/deploy_application.md (hunk 4 above)
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd-codecommit/deploy_application.md]: Used `bash` section

File: website/docs/automation/gitops/argocd-codecommit/cleanup.md (hunk 3 above)
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd-codecommit/cleanup.md:11]: I updated reset-environment, but those resources are created in argocd namespace, so I updated and use explicit clean up here.

**Comment by csantanapr** [comment]
@ybezsonov is the PR ready for another review, and addresses @niallthomson comments?

**Comment by ybezsonov** [comment]
> @ybezsonov is the PR ready for another review, and addresses @niallthomson comments?

@csantanapr Yes, I addressed comments and PR passed local make test module="automation/gitops"

**Comment by csantanapr** [comment]
@niallthomson when you have a chance could you merge this.

**Comment by ybezsonov** [comment]
Hi @niallthomson 
Could you please add e2e-tests label to this PR when it comes to merge?
It passes automated tests in automation/gitops, but it has changes to reset-environment and Dockerfile, so I would like to be sure that it will not  influence other modules.

File: website/docs/automation/gitops/argocd/update_application.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+```bash
+$ kubectl get deployment -n ui ui
+NAME   READY   UP-TO-DATE   AVAILABLE   AGE
+ui     3/3     3            3           3m33s
+$ kubectl get pod -n ui
+NAME                  READY   STATUS    RESTARTS   AGE
+ui-6d5bb7b95-hzmgp   1/1     Running   0          61s
+ui-6d5bb7b95-j28ww   1/1     Running   0          61s
+ui-6d5bb7b95-rjfxd   1/1     Running   0          3m34s
+```
```
Comment (candonov) [inline website/docs/automation/gitops/argocd/update_application.md:60]: Split in two, each command should go in its own code block.

**Review by candonov — COMMENTED** [review]
The Automation section start with GitOps. ArgoCD should be set to automatically sync the changes when pushed to git. @ybezsonov can you change the application to automatically sync?

File: website/docs/automation/gitops/argocd/update_application.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+Copy patch file to the Git repository directory:
+
+```bash
+$ cp /workspace/modules/automation/gitops/argocd/update-application/deployment-patch.yaml ~/environment/gitops/apps/deployment-patch.yaml
+```
+
+You can review planned changes in the file `/gitops/apps/deployment-patch.yaml`
+
+To apply the patch edit the file `/gitops/apps/kustomization.yaml` like in the example below:
```
[the remark this answers is not in the document]
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd/update_application.md]: Done.

File: website/docs/automation/gitops/argocd/update_application.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+```file
+automation/gitops/argocd/update-application/kustomization.yaml
+```
+
+You can execute commands to add necessary changes to the file `/gitops/apps/kustomization.yaml`:
```
[the remark this answers is not in the document]
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd/update_application.md]: Done.

File: website/docs/automation/gitops/argocd/update_application.md (hunk 6 above)
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd/update_application.md]: Done.

**Comment by ybezsonov** [comment]
> 

I improved Automation, added AutoSync to the application, described and provide example how to change default Argo CD Sync interval from 3 minutes to short one. Now a learner can get quick feedback on the changes.

File: environment/workspace/modules/automation/gitops/argocd/update-application/kustomization.yaml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+bases:
+- ../../../../../manifests/ui
```
Comment (candonov) [inline environment/workspace/modules/automation/gitops/argocd/update-application/kustomization.yaml:4]: Should this be `- ui`?

File: website/docs/automation/gitops/argocd/update_application.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+```bash
+$ cp /workspace/modules/automation/gitops/argocd/update-application/deployment-patch.yaml ~/environment/gitops/apps/deployment-patch.yaml
+```
+
+You can review planned changes in the file `apps/deployment-patch.yaml`
+
+To apply the patch edit the file `apps/kustomization.yaml` like in the example below:
+
+```file
+automation/gitops/argocd/update-application/kustomization.yaml
```
Comment (candonov) [inline website/docs/automation/gitops/argocd/update_application.md:25]:
```
$ cat apps/kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- ui
patches:
- deployment-patch.yaml
```
This relates to the `- ui` line being different in this diff.

File: website/docs/automation/gitops/argocd/deploy_application.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+You should now have all the resources related to the UI services deployed. To verify, run the following commands:
+
+```bash
+$ kubectl get deployment -n ui ui
+NAME   READY   UP-TO-DATE   AVAILABLE   AGE
+ui     1/1     1            1           61s
+$ kubectl get pod -n ui
+NAME                  READY   STATUS    RESTARTS   AGE
+ui-6d5bb7b95-rjfxd   1/1     Running   0          62s
+```
```
Comment (candonov) [inline website/docs/automation/gitops/argocd/deploy_application.md:103]: Can you split this one two please?

**Review by candonov — COMMENTED** [review]
Thank you for refactoring the sync. It is great. 
I left couple minor comments and it is ready to merge.

File: website/docs/automation/gitops/argocd/deploy_application.md (hunk 11 above)
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd/deploy_application.md]: Done.

File: environment/workspace/modules/automation/gitops/argocd/update-application/kustomization.yaml (hunk 9 above)
Comment (ybezsonov) [reply environment/workspace/modules/automation/gitops/argocd/update-application/kustomization.yaml:4]: This path was required to make kustomizing block show diffs. I changed approach and now it shows the same correct content like in the file which is used for deployment.

File: website/docs/automation/gitops/argocd/update_application.md (hunk 10 above)
Comment (ybezsonov) [reply website/docs/automation/gitops/argocd/update_application.md]: I changed approach and now it shows the same correct content like in the file which is used for deployment.

**Review by candonov — DISMISSED** [review]
/lgtm

**Comment by candonov** [comment]
@niallthomson I think you have to ✅ too. Also, the github actions e2e tests fail due to credentials. Do you know why the credentials are not working?

**Comment by csantanapr** [comment]
@niallthomson this one is ready for merging

**Comment by csantanapr** [comment]
@niallthomson would be good to know what's left on the PR to get it merge, please let @ybezsonov know

**Comment by niallthomson** [comment]
Thanks @ybezsonov for the content contribution!

**Pull request merged**